waku/v2/utils/pagination.nim

```diff
@@ -33,33 +33,28 @@ proc cmp*(x, y: Index): int =
   ## 
   ## Default sorting order priority is:
   ## 1. senderTimestamp
-  ## 2. receiverTimestamp (a fallback only if senderTimestamp unset and all other fields unequal)
+  ## 2. receiverTimestamp (a fallback only if senderTimestamp unset on either side, and all other fields unequal)
   ## 3. message digest
   ## 4. pubsubTopic
   
+  if x == y:
+    # Quick exit ensures receiver time does not affect index equality
+    return 0
+  
   # Timestamp has a higher priority for comparison
-  let timecmp = cmp(x.senderTime, y.senderTime)
+  var timecmp: int
+  if (x.senderTime == 0 or y.senderTime == 0):
+    # if either side does not have a senderTime set, we fall back to receiverTime comparison
+    timecmp = cmp(x.receiverTime, y.receiverTime)
```

Comment:
Re https://github.com/status-im/nim-waku/pull/864#discussion_r813805368
I'd say let's only fall back to `receivertime` for messages that don't have `senderTime`. I mean if x has sender time but y does not have, then compare x.senderTime with y.receiverTime. In the current implementation, a message with a valid `senderTime` sometimes gets compared with its `senderTime` and sometimes with its `receiverTime` which causes a non-deterministic behavior in the sorting output.

Reply:
Right, I had it like this and abandoned this approach because I thought comparing `receiverTime` only with `receiverTime` would be somewhat more accurate. However, you're right, I can now think of ways in which this may lead to (serious) inconsistencies in the underlying data structure. Have changed to suggested approach in [edfb68b](https://github.com/status-im/nim-waku/pull/864/commits/edfb68bdcde8bb2385210e0876158b2b7c1c6466) and modified unit tests to cover this case.